--- a/net/minecraft/block/entity/CommandBlockBlockEntity.java
+++ b/net/minecraft/block/entity/CommandBlockBlockEntity.java
@@ -23,6 +23,13 @@
     private boolean conditionMet;
     private boolean needsUpdatePacket;
     private final CommandBlockExecutor commandExecutor = new CommandBlockExecutor() {
+        // CraftBukkit start
+        @Override
+        public org.bukkit.command.CommandSender getBukkitSender(ServerCommandSource wrapper) {
+            return new org.bukkit.craftbukkit.command.CraftBlockCommandSender(wrapper, CommandBlockBlockEntity.this);
+        }
+        // CraftBukkit end
+
         public void setCommand(String s) {
             super.setCommand(s);
             CommandBlockBlockEntity.this.markDirty();
