--- a/net/minecraft/block/entity/LockableContainerBlockEntity.java
+++ b/net/minecraft/block/entity/LockableContainerBlockEntity.java
@@ -17,7 +17,7 @@
 
 public abstract class LockableContainerBlockEntity extends BlockEntity implements Inventory, NamedScreenHandlerFactory, Nameable {
 
-    private ContainerLock lock;
+    public ContainerLock lock; // Loom - public access
     private Text customName;
 
     protected LockableContainerBlockEntity(BlockEntityType<?> blockentitytype) {
@@ -83,4 +83,13 @@
     }
 
     protected abstract ScreenHandler createScreenHandler(int i, PlayerInventory arg);
+
+
+    // CraftBukkit start
+    @Override
+    public org.bukkit.Location getLocation() {
+        if (world == null) return null;
+        return new org.bukkit.Location(world.getCraftWorld(), pos.getX(), pos.getY(), pos.getZ());
+    }
+    // CraftBukkit end
 }
