--- a/net/minecraft/block/entity/SmokerBlockEntity.java
+++ b/net/minecraft/block/entity/SmokerBlockEntity.java
@@ -25,4 +25,14 @@
     protected ScreenHandler createScreenHandler(int i, PlayerInventory playerinventory) {
         return new SmokerScreenHandler(i, playerinventory, this, this.propertyDelegate);
     }
+
+    @Override
+    public int getMaxCountPerStack() {
+        return this.getMaxStackSize();
+    }
+
+    @Override
+    public void setMaxCountPerStack(int size) {
+        this.setMaxStackSize(size);
+    }
 }
