--- a/net/minecraft/block/sapling/LargeTreeSaplingGenerator.java
+++ b/net/minecraft/block/sapling/LargeTreeSaplingGenerator.java
@@ -38,6 +38,7 @@
             return false;
         } else {
             ((TreeFeatureConfig) configuredfeature.config).ignoreFluidCheck();
+            this.setTreeType(configuredfeature); // CraftBukkit
             BlockState blockstate1 = Blocks.AIR.getDefaultState();
 
             serverworld.setBlockState(blockpos.add(i, 0, j), blockstate1, 4);
