--- a/net/minecraft/block/sapling/SaplingGenerator.java
+++ b/net/minecraft/block/sapling/SaplingGenerator.java
@@ -12,6 +12,12 @@
 import net.minecraft.world.gen.feature.ConfiguredFeature;
 import net.minecraft.world.gen.feature.TreeFeatureConfig;
 
+// CraftBukkit start
+import org.bukkit.TreeType;
+import net.minecraft.world.gen.feature.DefaultBiomeFeatures;
+import net.minecraft.block.SaplingBlock;
+// CraftBukkit end
+
 public abstract class SaplingGenerator {
 
     public SaplingGenerator() {}
@@ -25,6 +31,7 @@
         if (configuredfeature == null) {
             return false;
         } else {
+            setTreeType(configuredfeature); // CraftBukkit
             serverworld.setBlockState(blockpos, Blocks.AIR.getDefaultState(), 4);
             ((TreeFeatureConfig) configuredfeature.config).ignoreFluidCheck();
             if (configuredfeature.generate(serverworld, serverworld.getStructureAccessor(), chunkgenerator, random, blockpos)) {
@@ -51,4 +58,46 @@
 
         return true;
     }
+
+    // CraftBukkit start
+    protected void setTreeType(ConfiguredFeature<?, ?> worldgentreeabstract) {
+        if (worldgentreeabstract.config == DefaultBiomeFeatures.OAK_TREE_CONFIG || worldgentreeabstract.config == DefaultBiomeFeatures.OAK_TREE_WITH_MORE_BEEHIVES_CONFIG) {
+            SaplingBlock.treeType = TreeType.TREE;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.HUGE_RED_MUSHROOM_CONFIG) {
+            SaplingBlock.treeType = TreeType.RED_MUSHROOM;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.HUGE_BROWN_MUSHROOM_CONFIG) {
+            SaplingBlock.treeType = TreeType.BROWN_MUSHROOM;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.JUNGLE_TREE_CONFIG) {
+            SaplingBlock.treeType = TreeType.COCOA_TREE;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.JUNGLE_SAPLING_TREE_CONFIG) {
+            SaplingBlock.treeType = TreeType.SMALL_JUNGLE;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.PINE_TREE_CONFIG) {
+            SaplingBlock.treeType = TreeType.TALL_REDWOOD;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.SPRUCE_TREE_CONFIG) {
+            SaplingBlock.treeType = TreeType.REDWOOD;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.ACACIA_TREE_CONFIG) {
+            SaplingBlock.treeType = TreeType.ACACIA;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.BIRCH_TREE_CONFIG || worldgentreeabstract.config == DefaultBiomeFeatures.BIRCH_TREE_WITH_MORE_BEEHIVES_CONFIG) {
+            SaplingBlock.treeType = TreeType.BIRCH;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.LARGE_BIRCH_TREE_CONFIG) {
+            SaplingBlock.treeType = TreeType.TALL_BIRCH;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.SWAMP_TREE_CONFIG) {
+            SaplingBlock.treeType = TreeType.SWAMP;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.FANCY_TREE_CONFIG || worldgentreeabstract.config == DefaultBiomeFeatures.FANCY_TREE_WITH_MORE_BEEHIVES_CONFIG) {
+            SaplingBlock.treeType = TreeType.BIG_TREE;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.JUNGLE_GROUND_BUSH_CONFIG) {
+            SaplingBlock.treeType = TreeType.JUNGLE_BUSH;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.DARK_OAK_TREE_CONFIG) {
+            SaplingBlock.treeType = TreeType.DARK_OAK;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.MEGA_SPRUCE_TREE_CONFIG) {
+            SaplingBlock.treeType = TreeType.MEGA_REDWOOD;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.MEGA_PINE_TREE_CONFIG) {
+            SaplingBlock.treeType = TreeType.MEGA_REDWOOD;
+        } else if (worldgentreeabstract.config == DefaultBiomeFeatures.MEGA_JUNGLE_TREE_CONFIG) {
+            SaplingBlock.treeType = TreeType.JUNGLE;
+        } else {
+            throw new IllegalArgumentException("Unknown tree generator " + worldgentreeabstract);
+        }
+    }
+    // CraftBukkit end
 }
