--- a/net/minecraft/screen/CartographyTableScreenHandler.java
+++ b/net/minecraft/screen/CartographyTableScreenHandler.java
@@ -40,6 +40,13 @@
                 CartographyTableScreenHandler.this.onContentChanged(this);
                 super.markDirty();
             }
+
+            // CraftBukkit start
+            @Override
+            public Location getLocation() {
+                return context.getLocation();
+            }
+            // CraftBukkit end
         };
         this.context = screenhandlercontext;
         this.addSlot(new Slot(this.inventory, 0, 15, 15) {
@@ -111,9 +118,11 @@
             this.addSlot(new Slot(playerinventory, j, 8 + j * 18, 142));
         }
 
+        player = (Player) playerinventory.player.getBukkitEntity(); // CraftBukkit
     }
 
     public boolean canUse(PlayerEntity playerentity) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return canUse(this.context, playerentity, Blocks.CARTOGRAPHY_TABLE);
     }
 
