--- a/net/minecraft/screen/LoomScreenHandler.java
+++ b/net/minecraft/screen/LoomScreenHandler.java
@@ -45,12 +45,26 @@
                 LoomScreenHandler.this.onContentChanged(this);
                 LoomScreenHandler.this.inventoryChangeListener.run();
             }
+
+            // CraftBukkit start
+            @Override
+            public Location getLocation() {
+                return context.getLocation();
+            }
+            // CraftBukkit end
         };
         this.output = new SimpleInventory(1) {
             public void markDirty() {
                 super.markDirty();
                 LoomScreenHandler.this.inventoryChangeListener.run();
             }
+
+            // CraftBukkit start
+            @Override
+            public Location getLocation() {
+                return context.getLocation();
+            }
+            // CraftBukkit end
         };
         this.context = screenhandlercontext;
         this.bannerSlot = this.addSlot(new Slot(this.input, 0, 13, 26) {
@@ -106,9 +120,11 @@
         }
 
         this.addProperty(this.selectedPattern);
+        player = (Player) playerinventory.player.getBukkitEntity(); // CraftBukkit
     }
 
     public boolean canUse(PlayerEntity playerentity) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return canUse(this.context, playerentity, Blocks.LOOM);
     }
 
@@ -223,6 +239,11 @@
 
                 if (compoundtag.contains("Patterns", 9)) {
                     listtag = compoundtag.getList("Patterns", 10);
+                    // CraftBukkit start
+                    while (listtag.size() > 20) {
+                        listtag.remove(20);
+                    }
+                    // CraftBukkit end
                 } else {
                     listtag = new ListTag();
                     compoundtag.put("Patterns", listtag);
