--- a/net/minecraft/screen/SmithingScreenHandler.java
+++ b/net/minecraft/screen/SmithingScreenHandler.java
@@ -11,6 +11,9 @@
 import net.minecraft.recipe.SmithingRecipe;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.inventory.CraftInventory;
+import org.bukkit.craftbukkit.inventory.CraftInventorySmithing;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public class SmithingScreenHandler extends ForgingScreenHandler {
 
@@ -18,6 +21,20 @@
     @Nullable
     private SmithingRecipe field_25386;
     private final List<SmithingRecipe> field_25668;
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity;
+
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventory inventory = new CraftInventorySmithing(this.input, this.output);
+        bukkitEntity = new CraftInventoryView(this.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
 
     public SmithingScreenHandler(int i, PlayerInventory playerinventory) {
         this(i, playerinventory, ScreenHandlerContext.EMPTY);
