--- a/net/minecraft/village/TradeOffer.java
+++ b/net/minecraft/village/TradeOffer.java
@@ -4,6 +4,7 @@
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.NbtHelper;
 import net.minecraft.util.math.MathHelper;
+import org.bukkit.craftbukkit.inventory.CraftMerchantRecipe; // CraftBukkit
 
 public class TradeOffer {
 
@@ -18,6 +19,19 @@
     private float priceMultiplier;
     private int traderExperience;
 
+    // CraftBukkit start
+    private CraftMerchantRecipe bukkitHandle;
+
+    public CraftMerchantRecipe asBukkit() {
+        return (bukkitHandle == null) ? bukkitHandle = new CraftMerchantRecipe(this) : bukkitHandle;
+    }
+
+    public TradeOffer(ItemStack itemstack, ItemStack itemstack1, ItemStack itemstack2, int uses, int maxUses, int experience, float priceMultiplier, CraftMerchantRecipe bukkit) {
+        this(itemstack, itemstack1, itemstack2, uses, maxUses, experience, priceMultiplier);
+        this.bukkitHandle = bukkit;
+    }
+    // CraftBukkit end
+
     public TradeOffer(CompoundTag compoundtag) {
         this.rewardingPlayerExperience = true;
         this.traderExperience = 1;
@@ -78,6 +92,7 @@
 
     public ItemStack getAdjustedFirstBuyItem() {
         int i = this.firstBuyItem.getCount();
+        if (i <= 0) return ItemStack.EMPTY; // CraftBukkit - SPIGOT-5476
         ItemStack itemstack = this.firstBuyItem.copy();
         int j = Math.max(0, MathHelper.floor((float) (i * this.demandBonus) * this.priceMultiplier));
 
