--- a/net/minecraft/village/raid/Raid.java
+++ b/net/minecraft/village/raid/Raid.java
@@ -67,21 +67,21 @@
     private static final Text DEFEAT_TITLE = Raid.EVENT_TEXT.shallowCopy().append(" - ").append((Text) Raid.DEFEAT_SUFFIX_TEXT);
     private final Map<Integer, RaiderEntity> waveToCaptain = Maps.newHashMap();
     private final Map<Integer, Set<RaiderEntity>> waveToRaiders = Maps.newHashMap();
-    private final Set<UUID> heroesOfTheVillage = Sets.newHashSet();
-    private long ticksActive;
+    public final Set<UUID> heroesOfTheVillage = Sets.newHashSet(); // Loom - change access
+    public long ticksActive; // Loom - change access
     private BlockPos center;
     private final ServerWorld world;
     private boolean started;
     private final int id;
     private float totalHealth;
-    private int badOmenLevel;
+    public int badOmenLevel; // Loom - change access
     private boolean active;
     private int wavesSpawned;
     private final ServerBossBar bar;
     private int postRaidTicks;
     private int preRaidTicks;
     private final Random random;
-    private final int waveCount;
+    public final int waveCount; // Loom - public access
     private Raid.Status status;
     private int finishCooldown;
     private Optional<BlockPos> preCalculatedRavagerSpawnLocation;
@@ -227,6 +227,7 @@
 
                 this.active = this.world.isChunkLoaded(this.center);
                 if (this.world.getDifficulty() == Difficulty.PEACEFUL) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.PEACE); // CraftBukkit
                     this.invalidate();
                     return;
                 }
@@ -246,13 +247,16 @@
                 if (!this.world.isNearOccupiedPointOfInterest(this.center)) {
                     if (this.wavesSpawned > 0) {
                         this.status = Raid.Status.LOSS;
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidFinishEvent(this, new java.util.ArrayList<>()); // CraftBukkit
                     } else {
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.NOT_IN_VILLAGE); // CraftBukkit
                         this.invalidate();
                     }
                 }
 
                 ++this.ticksActive;
                 if (this.ticksActive >= 48000L) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.TIMEOUT); // CraftBukkit
                     this.invalidate();
                     return;
                 }
@@ -326,6 +330,7 @@
                     }
 
                     if (j > 3) {
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.UNSPAWNABLE);  // CraftBukkit
                         this.invalidate();
                         break;
                     }
@@ -338,6 +343,7 @@
                         this.status = Raid.Status.VICTORY;
                         Iterator iterator = this.heroesOfTheVillage.iterator();
 
+                        List<org.bukkit.entity.Player> winners = new java.util.ArrayList<>(); // CraftBukkit
                         while (iterator.hasNext()) {
                             UUID uuid = (UUID) iterator.next();
                             Entity entity = this.world.getEntity(uuid);
@@ -351,9 +357,11 @@
 
                                     serverplayerentity.incrementStat(Stats.RAID_WIN);
                                     Criteria.HERO_OF_THE_VILLAGE.trigger(serverplayerentity);
+                                    winners.add(serverplayerentity.getBukkitEntity()); // CraftBukkit
                                 }
                             }
                         }
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidFinishEvent(this, winners); // CraftBukkit
                     }
                 }
 
@@ -361,6 +369,7 @@
             } else if (this.isFinished()) {
                 ++this.finishCooldown;
                 if (this.finishCooldown >= 600) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, org.bukkit.event.raid.RaidStopEvent.Reason.FINISHED); // CraftBukkit
                     this.invalidate();
                     return;
                 }
@@ -495,6 +504,10 @@
         Raid.Member[] araid_member = Raid.Member.VALUES;
         int j = araid_member.length;
 
+        // CraftBukkit start
+        RaiderEntity leader = null;
+        List<RaiderEntity> raiders = new java.util.ArrayList<>();
+        // CraftBukkit end
         for (int k = 0; k < j; ++k) {
             Raid.Member raid_member = araid_member[k];
             int l = this.getCount(raid_member, i, flag1) + this.getBonusCount(raid_member, this.random, i, localdifficulty, flag1);
@@ -507,9 +520,11 @@
                     raiderentity.setPatrolLeader(true);
                     this.setWaveCaptain(i, raiderentity);
                     flag = true;
+                    leader = raiderentity; // CraftBukkit
                 }
 
                 this.addRaider(i, raiderentity, blockpos, false);
+                raiders.add(raiderentity); // CraftBukkit
                 if (raid_member.type == EntityType.RAVAGER) {
                     RaiderEntity raiderentity1 = null;
 
@@ -528,6 +543,7 @@
                         this.addRaider(i, raiderentity1, blockpos, false);
                         raiderentity1.refreshPositionAndAngles(blockpos, 0.0F, 0.0F);
                         raiderentity1.startRiding(raiderentity);
+                        raiders.add(raiderentity); // CraftBukkit // TODO should be raiderentity1?
                     }
                 }
             }
@@ -537,6 +553,7 @@
         ++this.wavesSpawned;
         this.updateBar();
         this.markDirty();
+        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidSpawnWaveEvent(this, leader, raiders); // CraftBukkit
     }
 
     public void addRaider(int i, RaiderEntity raiderentity, @Nullable BlockPos blockpos, boolean flag) {
@@ -552,7 +569,7 @@
                 raiderentity.initialize(this.world, this.world.getLocalDifficulty(blockpos), SpawnReason.EVENT, (EntityData) null, (CompoundTag) null);
                 raiderentity.addBonusForWave(i, false);
                 raiderentity.setOnGround(true);
-                this.world.spawnEntity(raiderentity);
+                this.world.spawnEntity(raiderentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.RAID); // CraftBukkit
             }
         }
 
@@ -811,6 +828,16 @@
         }
     }
 
+    // CraftBukkit start
+    public boolean isInProgress() {
+        return this.status == Status.ONGOING;
+    }
+
+    public java.util.Collection<RaiderEntity> getRaiders() {
+        return this.waveToRaiders.values().stream().flatMap(Set::stream).collect(java.util.stream.Collectors.toSet());
+    }
+    // CraftBukkit end
+
     static enum Status {
 
         ONGOING, VICTORY, LOSS, STOPPED;
