--- a/net/minecraft/entity/projectile/ExplosiveProjectileEntity.java
+++ b/net/minecraft/entity/projectile/ExplosiveProjectileEntity.java
@@ -15,12 +15,15 @@
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.RayTraceContext;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public abstract class ExplosiveProjectileEntity extends ProjectileEntity {
 
     public double posX;
     public double posY;
     public double posZ;
+    public float bukkitYield = 1; // CraftBukkit
+    public boolean isIncendiary = true; // CraftBukkit
 
     protected ExplosiveProjectileEntity(EntityType<? extends ExplosiveProjectileEntity> entitytype, World world) {
         super(entitytype, world);
@@ -30,6 +33,12 @@
         this(entitytype, world);
         this.refreshPositionAndAngles(d0, d1, d2, this.yaw, this.pitch);
         this.refreshPosition();
+        // CraftBukkit start - Added setDirection method
+        this.setDirection(d3, d4, d5);
+    }
+
+    public void setDirection(double d3, double d4, double d5) {
+        // CraftBukkit end
         double d6 = (double) MathHelper.sqrt(d3 * d3 + d4 * d4 + d5 * d5);
 
         if (d6 != 0.0D) {
@@ -63,6 +72,12 @@
 
             if (hitresult.getType() != HitResult.Type.MISS) {
                 this.onCollision(hitresult);
+
+                // CraftBukkit start - Fire ProjectileHitEvent
+                if (this.removed) {
+                    CraftEventFactory.callProjectileHitEvent(this, hitresult);
+                }
+                // CraftBukkit end
             }
 
             Vec3d vec3d = this.getVelocity();
@@ -140,6 +155,11 @@
             Entity entity = damagesource.getAttacker();
 
             if (entity != null) {
+                // CraftBukkit start
+                if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, damagesource, f)) {
+                    return false;
+                }
+                // CraftBukkit end
                 Vec3d vec3d = entity.getRotationVector();
 
                 this.setVelocity(vec3d);
