--- a/net/minecraft/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/entity/projectile/FireworkRocketEntity.java
@@ -28,14 +28,15 @@
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.RayTraceContext;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class FireworkRocketEntity extends ProjectileEntity {
 
-    private static final TrackedData<ItemStack> ITEM = DataTracker.registerData(FireworkRocketEntity.class, TrackedDataHandlerRegistry.ITEM_STACK);
+    public static final TrackedData<ItemStack> ITEM = DataTracker.registerData(FireworkRocketEntity.class, TrackedDataHandlerRegistry.ITEM_STACK); // Loom - public access
     private static final TrackedData<OptionalInt> SHOOTER_ENTITY_ID = DataTracker.registerData(FireworkRocketEntity.class, TrackedDataHandlerRegistry.FIREWORK_DATA);
-    private static final TrackedData<Boolean> SHOT_AT_ANGLE = DataTracker.registerData(FireworkRocketEntity.class, TrackedDataHandlerRegistry.BOOLEAN);
+    public static final TrackedData<Boolean> SHOT_AT_ANGLE = DataTracker.registerData(FireworkRocketEntity.class, TrackedDataHandlerRegistry.BOOLEAN); // Loom - public access
     private int life;
-    private int lifeTime;
+    public int lifeTime; // Loom - public access
     private LivingEntity shooter;
 
     public FireworkRocketEntity(EntityType<? extends FireworkRocketEntity> entitytype, World world) {
@@ -141,7 +142,11 @@
         }
 
         if (!this.world.isClient && this.life > this.lifeTime) {
-            this.explodeAndRemove();
+            // CraftBukkit start
+            if (!CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+                this.explodeAndRemove();
+            }
+            // CraftBukkit end
         }
 
     }
@@ -155,7 +160,11 @@
     protected void onEntityHit(EntityHitResult entityhitresult) {
         super.onEntityHit(entityhitresult);
         if (!this.world.isClient) {
-            this.explodeAndRemove();
+            // CraftBukkit start
+            if (!CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+                this.explodeAndRemove();
+            }
+            // CraftBukkit end
         }
     }
 
@@ -190,7 +199,9 @@
 
         if (f > 0.0F) {
             if (this.shooter != null) {
+                CraftEventFactory.entityDamage = this; // CraftBukkit
                 this.shooter.damage(DamageSource.firework(this, this.getOwner()), 5.0F + (float) (listtag.size() * 2));
+                CraftEventFactory.entityDamage = null; // CraftBukkit
             }
 
             double d0 = 5.0D;
@@ -217,7 +228,9 @@
                     if (flag) {
                         float f1 = f * (float) Math.sqrt((5.0D - (double) this.distanceTo(livingentity)) / 5.0D);
 
+                        CraftEventFactory.entityDamage = this; // CraftBukkit
                         livingentity.damage(DamageSource.firework(this, this.getOwner()), f1);
+                        CraftEventFactory.entityDamage = null; // CraftBukkit
                     }
                 }
             }
