--- a/net/minecraft/block/dispenser/ProjectileDispenserBehavior.java
+++ b/net/minecraft/block/dispenser/ProjectileDispenserBehavior.java
@@ -1,12 +1,18 @@
 package net.minecraft.block.dispenser;
 
 import net.minecraft.block.DispenserBlock;
+import net.minecraft.block.entity.DispenserBlockEntity;
+import net.minecraft.entity.Entity;
 import net.minecraft.entity.projectile.ProjectileEntity;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.math.BlockPointer;
+import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.Direction;
 import net.minecraft.util.math.Position;
 import net.minecraft.world.World;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
 
 public abstract class ProjectileDispenserBehavior extends ItemDispenserBehavior {
 
@@ -18,9 +24,39 @@
         Direction direction = (Direction) blockpointer.getBlockState().get(DispenserBlock.FACING);
         ProjectileEntity projectileentity = this.createProjectile(world, position, itemstack);
 
-        projectileentity.setVelocity((double) direction.getOffsetX(), (double) ((float) direction.getOffsetY() + 0.1F), (double) direction.getOffsetZ(), this.getForce(), this.getVariation());
+        //projectileentity.setVelocity((double) direction.getOffsetX(), (double) ((float) direction.getOffsetY() + 0.1F), (double) direction.getOffsetZ(), this.getForce(), this.getVariation());
+        // CraftBukkit start
+        ItemStack itemstack1 = itemstack.split(1);
+        BlockPos blockpos = blockpointer.getBlockPos();
+        org.bukkit.block.Block block = world.getCraftWorld().getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ());
+        CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+
+        BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector((double) direction.getOffsetX(), (float) direction.getOffsetY() + 0.1F, direction.getOffsetZ()));
+        if (!DispenserBlock.eventFired) {
+            Bukkit.getPluginManager().callEvent(event);
+        }
+
+        if (event.isCancelled()) {
+            itemstack.increment(1);
+            return itemstack;
+        }
+
+        if (!event.getItem().equals(craftItem)) {
+            itemstack.increment(1);
+            // Chain to handler for new item
+            ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+            DispenserBehavior idispensebehavior = DispenserBlock.BEHAVIORS.get(eventStack.getItem());
+            if (idispensebehavior != DispenserBehavior.NOOP && idispensebehavior != this) {
+                idispensebehavior.dispense(blockpointer, eventStack);
+                return itemstack;
+            }
+        }
+
+        projectileentity.setVelocity(event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), this.getForce(), this.getVariation());
+        projectileentity.projectileSource = new org.bukkit.craftbukkit.projectiles.CraftBlockProjectileSource((DispenserBlockEntity) blockpointer.getBlockEntity());
+        // CraftBukkit end
         world.spawnEntity(projectileentity);
-        itemstack.decrement(1);
+        // itemstack.decrement(1); // CraftBukkit - Handled during event processing
         return itemstack;
     }
 
