--- a/net/minecraft/entity/player/ItemCooldownManager.java
+++ b/net/minecraft/entity/player/ItemCooldownManager.java
@@ -8,8 +8,8 @@
 
 public class ItemCooldownManager {
 
-    private final Map<Item, ItemCooldownManager.Entry> entries = Maps.newHashMap();
-    private int tick;
+    public final Map<Item, ItemCooldownManager.Entry> entries = Maps.newHashMap(); // Loom - public access
+    public int tick; // Loom - public access
 
     public ItemCooldownManager() {}
 
@@ -56,10 +56,10 @@
 
     protected void onCooldownUpdate(Item item) {}
 
-    class Entry {
+    public class Entry { // Loom - public access
 
         private final int startTick;
-        private final int endTick;
+        public final int endTick; // Loom - public access
 
         private Entry(int i, int j) {
             this.startTick = i;
