--- a/net/minecraft/entity/passive/CowEntity.java
+++ b/net/minecraft/entity/passive/CowEntity.java
@@ -21,12 +21,16 @@
 import net.minecraft.item.ItemUsage;
 import net.minecraft.item.Items;
 import net.minecraft.recipe.Ingredient;
+import net.minecraft.server.world.ServerWorld;
 import net.minecraft.sound.SoundEvent;
 import net.minecraft.sound.SoundEvents;
 import net.minecraft.util.ActionResult;
 import net.minecraft.util.Hand;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerBucketFillEvent;
 
 public class CowEntity extends AnimalEntity {
 
@@ -49,6 +53,7 @@
         return MobEntity.createMobAttributes().add(EntityAttributes.GENERIC_MAX_HEALTH, 10.0D).add(EntityAttributes.GENERIC_MOVEMENT_SPEED, 0.20000000298023224D);
     }
 
+
     protected SoundEvent getAmbientSound() {
         return SoundEvents.ENTITY_COW_AMBIENT;
     }
@@ -73,8 +78,14 @@
         ItemStack itemstack = playerentity.getStackInHand(hand);
 
         if (itemstack.getItem() == Items.BUCKET && !this.isBaby()) {
+            // CraftBukkit start - Got milk?
+            PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent((ServerWorld) playerentity.world, playerentity, this.getBlockPos(), this.getBlockPos(), null, itemstack, Items.MILK_BUCKET);
+            if (event.isCancelled()) {
+                return ActionResult.PASS;
+            }
+            // CraftBukkit end
             playerentity.playSound(SoundEvents.ENTITY_COW_MILK, 1.0F, 1.0F);
-            ItemStack itemstack1 = ItemUsage.method_30012(itemstack, playerentity, Items.MILK_BUCKET.getStackForRender());
+            ItemStack itemstack1 = ItemUsage.method_30012(itemstack, playerentity, CraftItemStack.asNMSCopy(event.getItemStack())); // CraftBukkit
 
             playerentity.setStackInHand(hand, itemstack1);
             return ActionResult.success(this.world.isClient);
@@ -91,7 +102,9 @@
         return this.isBaby() ? entitydimensions.height * 0.95F : 1.3F;
     }
 
+    /*
     public PassiveEntity createChild(PassiveEntity passiveentity) {
         return this.createChild(passiveentity);
     }
+    */ // Loom - decompile error
 }
