--- a/net/minecraft/entity/passive/PigEntity.java
+++ b/net/minecraft/entity/passive/PigEntity.java
@@ -46,12 +46,16 @@
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.World;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+// CraftBukkit end
+
 public class PigEntity extends AnimalEntity implements ItemSteerable, Saddleable {
 
     private static final TrackedData<Boolean> SADDLED = DataTracker.registerData(PigEntity.class, TrackedDataHandlerRegistry.BOOLEAN);
     private static final TrackedData<Integer> BOOST_TIME = DataTracker.registerData(PigEntity.class, TrackedDataHandlerRegistry.INTEGER);
     private static final Ingredient BREEDING_INGREDIENT = Ingredient.ofItems(Items.CARROT, Items.POTATO, Items.BEETROOT);
-    private final SaddledComponent saddledComponent;
+    public final SaddledComponent saddledComponent; // Loom - public access
 
     public PigEntity(EntityType<? extends PigEntity> entitytype, World world) {
         super(entitytype, world);
@@ -230,6 +234,13 @@
 
             zombifiedpiglinentity.setPersistent();
             this.world.spawnEntity(zombifiedpiglinentity);
+            // CraftBukkit start
+            if (CraftEventFactory.callPigZapEvent(this, lightningentity, zombifiedpiglinentity).isCancelled()) {
+                return;
+            }
+            // CraftBukkit - added a reason for spawning this creature
+            this.world.spawnEntity(zombifiedpiglinentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
+            // CraftBukkit end
             this.remove();
         } else {
             super.onStruckByLightning(lightningentity);
@@ -261,7 +272,5 @@
         return PigEntity.BREEDING_INGREDIENT.test(itemstack);
     }
 
-    public PassiveEntity createChild(PassiveEntity passiveentity) {
-        return this.createChild(passiveentity);
-    }
+    // Loom - decompile errors
 }
