--- a/net/minecraft/entity/projectile/thrown/ThrownItemEntity.java
+++ b/net/minecraft/entity/projectile/thrown/ThrownItemEntity.java
@@ -30,7 +30,7 @@
     public void setItem(ItemStack itemstack) {
         if (itemstack.getItem() != this.getDefaultItem() || itemstack.hasTag()) {
             this.getDataTracker().set(ThrownItemEntity.ITEM, Util.make(itemstack.copy(), (itemstack1) -> {
-                itemstack1.setCount(1);
+                if (!itemstack1.isEmpty()) itemstack1.setCount(1); // CraftBukkit
             }));
         }
 
@@ -38,7 +38,13 @@
 
     protected abstract Item getDefaultItem();
 
-    protected ItemStack getItem() {
+    // CraftBukkit start
+    public Item getDefaultItemPublic() {
+        return getDefaultItem();
+    }
+    // CraftBukkit end
+
+    public ItemStack getItem() { // Loom - public access
         return (ItemStack) this.getDataTracker().get(ThrownItemEntity.ITEM);
     }
 
