--- a/net/minecraft/world/border/WorldBorder.java
+++ b/net/minecraft/world/border/WorldBorder.java
@@ -6,6 +6,7 @@
 import java.util.List;
 import net.minecraft.entity.Entity;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.server.world.ServerWorld;
 import net.minecraft.util.Util;
 import net.minecraft.util.function.BooleanBiFunction;
 import net.minecraft.util.math.BlockPos;
@@ -27,6 +28,7 @@
     private int maxWorldBorderRadius = 29999984;
     private WorldBorder.Area area = new WorldBorder.StaticArea(6.0E7D);
     public static final WorldBorder.Properties DEFAULT_BORDER = new WorldBorder.Properties(0.0D, 0.0D, 0.2D, 5.0D, 5, 15, 6.0E7D, 0L, 0.0D);
+    public ServerWorld world; // CraftBukkit
 
     public WorldBorder() {}
 
@@ -140,6 +142,7 @@
     }
 
     public void addListener(WorldBorderListener worldborderlistener) {
+        if (listeners.contains(worldborderlistener)) return; // CraftBukkit
         this.listeners.add(worldborderlistener);
     }
 
