--- a/net/minecraft/world/updater/WorldUpdater.java
+++ b/net/minecraft/world/updater/WorldUpdater.java
@@ -27,6 +27,7 @@
 import net.minecraft.util.registry.RegistryKey;
 import net.minecraft.world.PersistentStateManager;
 import net.minecraft.world.World;
+import net.minecraft.world.dimension.DimensionType;
 import net.minecraft.world.level.storage.LevelStorage;
 import net.minecraft.world.storage.RegionFile;
 import net.minecraft.world.storage.VersionedChunkStorage;
@@ -37,7 +38,7 @@
 
     private static final Logger LOGGER = LogManager.getLogger();
     private static final ThreadFactory UPDATE_THREAD_FACTORY = (new ThreadFactoryBuilder()).setDaemon(true).build();
-    private final ImmutableSet<RegistryKey<World>> worlds;
+    private final ImmutableSet<RegistryKey<DimensionType>> worlds;
     private final boolean eraseCache;
     private final LevelStorage.Session session;
     private final Thread updateThread;
@@ -53,7 +54,7 @@
     private static final Pattern REGION_FILE_PATTERN = Pattern.compile("^r\.(-?[0-9]+)\.(-?[0-9]+)\.mca$");
     private final PersistentStateManager persistentStateManager;
 
-    public WorldUpdater(LevelStorage.Session levelstorage_session, DataFixer datafixer, ImmutableSet<RegistryKey<World>> immutableset, boolean flag) {
+    public WorldUpdater(LevelStorage.Session levelstorage_session, DataFixer datafixer, ImmutableSet<RegistryKey<DimensionType>> immutableset, boolean flag) { // CraftBukkit
         this.worlds = immutableset;
         this.eraseCache = flag;
         this.dataFixer = datafixer;
@@ -132,9 +133,7 @@
 
                             if (compoundtag != null) {
                                 int j = VersionedChunkStorage.getDataVersion(compoundtag);
-                                CompoundTag compoundtag1 = versionedchunkstorage.updateChunkTag(registrykey2, () -> {
-                                    return this.persistentStateManager;
-                                }, compoundtag);
+                                CompoundTag compoundtag1 = versionedchunkstorage.updateChunkTag(registrykey2, () -> this.persistentStateManager, compoundtag, chunkpos, null); // CraftBukkit
                                 CompoundTag compoundtag2 = compoundtag1.getCompound("Level");
                                 ChunkPos chunkpos1 = new ChunkPos(compoundtag2.getInt("xPos"), compoundtag2.getInt("zPos"));
 
