--- a/net/minecraft/network/packet/s2c/play/GameStateChangeS2CPacket.java
+++ b/net/minecraft/network/packet/s2c/play/GameStateChangeS2CPacket.java
@@ -46,9 +46,7 @@
         clientplaypacketlistener.onGameStateChange(this);
     }
 
-    public void apply(PacketListener packetlistener) {
-        this.apply((ClientPlayPacketListener) packetlistener);
-    }
+    // Loom - decompile errors
 
     public static class Reason {
 
