--- a/net/minecraft/network/packet/s2c/play/PlayerListHeaderS2CPacket.java
+++ b/net/minecraft/network/packet/s2c/play/PlayerListHeaderS2CPacket.java
@@ -9,8 +9,7 @@
 
 public class PlayerListHeaderS2CPacket implements Packet<ClientPlayPacketListener> {
 
-    private Text header;
-    private Text footer;
+    public Text header, footer; // Loom - public access
 
     public PlayerListHeaderS2CPacket() {}
 
@@ -27,8 +26,4 @@
     public void apply(ClientPlayPacketListener clientplaypacketlistener) {
         clientplaypacketlistener.onPlayerListHeader(this);
     }
-
-    public void apply(PacketListener packetlistener) {
-        this.apply((ClientPlayPacketListener) packetlistener);
-    }
 }
