--- a/net/minecraft/server/rcon/RconCommandOutput.java
+++ b/net/minecraft/server/rcon/RconCommandOutput.java
@@ -5,6 +5,7 @@
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.command.CommandOutput;
 import net.minecraft.server.command.ServerCommandSource;
+import net.minecraft.server.world.ServerWorld;
 import net.minecraft.text.LiteralText;
 import net.minecraft.text.Text;
 import net.minecraft.util.math.Vec2f;
@@ -49,4 +50,15 @@
     public boolean shouldBroadcastConsoleToOps() {
         return this.server.shouldBroadcastRconToOps();
     }
+
+    // CraftBukkit start - Send a String
+    public void sendMessage(String message) {
+        this.buffer.append(message);
+    }
+
+    @Override
+    public org.bukkit.command.CommandSender getBukkitSender(ServerCommandSource wrapper) {
+        return server.remoteConsole; // TODO patch
+    }
+    // CraftBukkit end
 }
