--- a/net/minecraft/advancement/PlayerAdvancementTracker.java
+++ b/net/minecraft/advancement/PlayerAdvancementTracker.java
@@ -48,6 +48,7 @@
 import net.minecraft.world.GameRules;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.event.player.PlayerAdvancementDoneEvent;
 
 public class PlayerAdvancementTracker {
 
@@ -179,7 +180,11 @@
                         Advancement advancement = serveradvancementloader.get((Identifier) entry.getKey());
 
                         if (advancement == null) {
-                            PlayerAdvancementTracker.LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", entry.getKey(), (Object) this.advancementFile);
+                            // CraftBukkit start
+                            if (entry.getKey().getNamespace().equals("minecraft")) {
+                                PlayerAdvancementTracker.LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", entry.getKey(), this.advancementFile);
+                            }
+                            // CraftBukkit end
                         } else {
                             this.initProgress(advancement, (AdvancementProgress) entry.getValue());
                         }
@@ -294,6 +299,7 @@
             this.progressUpdates.add(advancement);
             flag = true;
             if (!flag1 && advancementprogress.isDone()) {
+                this.owner.world.getServer().getPluginManager().callEvent(new PlayerAdvancementDoneEvent(this.owner.getBukkitEntity(), advancement.bukkit)); // CraftBukkit
                 advancement.getRewards().apply(this.owner);
                 if (advancement.getDisplay() != null && advancement.getDisplay().shouldAnnounceToChat() && this.owner.world.getGameRules().getBoolean(GameRules.ANNOUNCE_ADVANCEMENTS)) {
                     this.field_25325.broadcastChatMessage(new TranslatableText("chat.type.advancement." + advancement.getDisplay().getFrame().getId(), new Object[]{this.owner.getDisplayName(), advancement.toHoverableText()}), MessageType.SYSTEM, Util.NIL_UUID);
