--- a/net/minecraft/item/BlockItem.java
+++ b/net/minecraft/item/BlockItem.java
@@ -22,6 +22,13 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
+// CraftBukkit start
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.block.data.CraftBlockData;
+import org.bukkit.event.block.BlockCanBuildEvent;
+// CraftBukkit end
+
 public class BlockItem extends Item {
 
     @Deprecated
@@ -72,7 +79,7 @@
 
                     BlockSoundGroup blocksoundgroup = blockstate1.getSoundGroup();
 
-                    world.playSound(playerentity, blockpos, this.getPlaceSound(blockstate1), SoundCategory.BLOCKS, (blocksoundgroup.getVolume() + 1.0F) / 2.0F, blocksoundgroup.getPitch() * 0.8F);
+                    // world.playSound(playerentity, blockpos, this.getPlaceSound(blockstate1), SoundCategory.BLOCKS, (blocksoundgroup.getVolume() + 1.0F) / 2.0F, blocksoundgroup.getPitch() * 0.8F); // CraftBukkit
                     if (playerentity == null || !playerentity.abilities.creativeMode) {
                         itemstack.decrement(1);
                     }
@@ -109,6 +116,21 @@
 
         if (compoundtag != null) {
             CompoundTag compoundtag1 = compoundtag.getCompound("BlockStateTag");
+            // CraftBukkit start
+            blockstate1 = getBlockState(blockstate1, compoundtag1);
+        }
+
+        if (blockstate1 != blockstate) {
+            world.setBlockState(blockpos, blockstate1, 2);
+        }
+
+        return blockstate1;
+    }
+
+    public static BlockState getBlockState(BlockState blockstate, CompoundTag compoundtag1) {
+        BlockState blockstate1 = blockstate;
+        {
+            // CraftBukkit end
             StateManager<Block, BlockState> statemanager = blockstate.getBlock().getStateManager();
             Iterator iterator = compoundtag1.getKeys().iterator();
 
@@ -123,11 +145,6 @@
                 }
             }
         }
-
-        if (blockstate1 != blockstate) {
-            world.setBlockState(blockpos, blockstate1, 2);
-        }
-
         return blockstate1;
     }
 
@@ -141,7 +158,14 @@
         PlayerEntity playerentity = itemplacementcontext.getPlayer();
         ShapeContext shapecontext = playerentity == null ? ShapeContext.absent() : ShapeContext.of(playerentity);
 
-        return (!this.checkStatePlacement() || blockstate.canPlaceAt(itemplacementcontext.getWorld(), itemplacementcontext.getBlockPos())) && itemplacementcontext.getWorld().canPlace(blockstate, itemplacementcontext.getBlockPos(), shapecontext);
+        // CraftBukkit start - store default return
+        boolean defaultReturn = (!this.checkStatePlacement() || blockstate.canPlaceAt(itemplacementcontext.getWorld(), itemplacementcontext.getBlockPos())) && itemplacementcontext.getWorld().canPlace(blockstate, itemplacementcontext.getBlockPos(), shapecontext);
+        org.bukkit.entity.Player player = (itemplacementcontext.getPlayer() instanceof ServerPlayerEntity) ? (org.bukkit.entity.Player) itemplacementcontext.getPlayer().getBukkitEntity() : null;
+
+        BlockCanBuildEvent event = new BlockCanBuildEvent(CraftBlock.at(itemplacementcontext.getWorld(), itemplacementcontext.getBlockPos()), player, CraftBlockData.fromData(blockstate), defaultReturn);
+        Bukkit.getPluginManager().callEvent(event);
+        return event.isBuildable();
+        // CraftBukkit end
     }
 
     protected boolean checkStatePlacement() {
