--- a/net/minecraft/item/CrossbowItem.java
+++ b/net/minecraft/item/CrossbowItem.java
@@ -96,6 +96,12 @@
                 itemstack2 = itemstack1.copy();
             }
 
+            // CraftBukkit start - SPIGOT-4870, MC-150847
+            else if (itemstack1.isEmpty()) {
+                return false;
+            }
+            // CraftBukkit end
+
             if (!loadProjectile(livingentity, itemstack, itemstack1, k > 0, flag)) {
                 return false;
             }
@@ -218,11 +224,27 @@
                 vector3f.rotate(quaternion);
                 ((ProjectileEntity) object).setVelocity((double) vector3f.getX(), (double) vector3f.getY(), (double) vector3f.getZ(), f1, f2);
             }
+            // CraftBukkit start
+            org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityShootBowEvent(livingentity, itemstack, (Entity) object, f);
+            if (event.isCancelled()) {
+                event.getProjectile().remove();
+                return;
+            }
+            // CraftBukkit end
 
             itemstack.damage(flag1 ? 3 : 1, livingentity, (livingentity1) -> {
                 livingentity1.sendToolBreakStatus(hand);
             });
-            world.spawnEntity((Entity) object);
+            // CraftBukkit start
+            if (event.getProjectile() == ((Entity) object).getBukkitEntity()) {
+                if (!world.spawnEntity((Entity) object)) {
+                    if (livingentity instanceof ServerPlayerEntity) {
+                        ((ServerPlayerEntity) livingentity).getBukkitEntity().updateInventory();
+                    }
+                    return;
+                }
+            }
+            // CraftBukkit end
             world.playSound((PlayerEntity) null, livingentity.getX(), livingentity.getY(), livingentity.getZ(), SoundEvents.ITEM_CROSSBOW_SHOOT, SoundCategory.PLAYERS, 1.0F, f);
         }
     }
