--- a/net/minecraft/world/gen/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/gen/chunk/ChunkGenerator.java
@@ -60,7 +60,7 @@
         this.field_24748 = i;
     }
 
-    private void method_28509() {
+    public void method_28509() { // Loom - public access
         if (this.field_24749.isEmpty()) {
             StrongholdConfig strongholdconfig = this.config.getStronghold();
 
@@ -243,7 +243,15 @@
         while (iterator.hasNext()) {
             ConfiguredStructureFeature<?, ?> configuredstructurefeature = (ConfiguredStructureFeature) iterator.next();
 
-            this.method_28508(configuredstructurefeature, structureaccessor, chunk, structuremanager, i, chunkpos, biome);
+            // CraftBukkit start
+            if (configuredstructurefeature.field_24835 == StructureFeature.STRONGHOLD) {
+                synchronized (configuredstructurefeature) {
+                    this.method_28508(configuredstructurefeature, structureaccessor, chunk, structuremanager, i, chunkpos, biome);
+                }
+            } else {
+                this.method_28508(configuredstructurefeature, structureaccessor, chunk, structuremanager, i, chunkpos, biome);
+            }
+            // CraftBukkit end
         }
 
     }
