--- a/net/minecraft/recipe/BlastingRecipe.java
+++ b/net/minecraft/recipe/BlastingRecipe.java
@@ -3,6 +3,14 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.Identifier;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftBlastingRecipe;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.util.CraftNamespacedKey;
+import org.bukkit.inventory.Recipe;
+// CraftBukkit end
+
 public class BlastingRecipe extends AbstractCookingRecipe {
 
     public BlastingRecipe(Identifier identifier, String s, Ingredient ingredient, ItemStack itemstack, float f, int i) {
@@ -12,4 +20,16 @@
     public RecipeSerializer<?> getSerializer() {
         return RecipeSerializer.BLASTING;
     }
+
+    // CraftBukkit start
+    @Override
+    public Recipe toBukkitRecipe() {
+        CraftItemStack result = CraftItemStack.asCraftMirror(this.output);
+
+        CraftBlastingRecipe recipe = new CraftBlastingRecipe(CraftNamespacedKey.fromMinecraft(this.id), result, CraftRecipe.toBukkit(this.input), this.experience, this.cookTime);
+        recipe.setGroup(this.group);
+
+        return recipe;
+    }
+    // CraftBukkit end
 }
