--- a/net/minecraft/recipe/SmithingRecipe.java
+++ b/net/minecraft/recipe/SmithingRecipe.java
@@ -9,6 +9,13 @@
 import net.minecraft.util.JsonHelper;
 import net.minecraft.world.World;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.inventory.CraftSmithingRecipe;
+import org.bukkit.craftbukkit.util.CraftNamespacedKey;
+// CraftBukkit end
+
 public class SmithingRecipe implements Recipe<Inventory> {
 
     private final Ingredient base;
@@ -84,16 +91,19 @@
             packetbytebuf.writeItemStack(smithingrecipe.result);
         }
 
-        public void write(PacketByteBuf packetbytebuf, Recipe recipe) {
-            this.write(packetbytebuf, (SmithingRecipe) recipe);
-        }
+        // Loom - decompile error
+    }
 
-        public Recipe read(Identifier identifier, PacketByteBuf packetbytebuf) {
-            return this.read(identifier, packetbytebuf);
-        }
 
-        public Recipe read(Identifier identifier, JsonObject jsonobject) {
-            return this.read(identifier, jsonobject);
-        }
+    // CraftBukkit start
+    @Override
+    public org.bukkit.inventory.Recipe toBukkitRecipe() {
+        CraftItemStack result = CraftItemStack.asCraftMirror(this.getOutput());
+
+        CraftSmithingRecipe recipe = new CraftSmithingRecipe(CraftNamespacedKey.fromMinecraft(this.id), result, CraftRecipe.toBukkit(this.base), CraftRecipe.toBukkit(this.addition));
+
+        return recipe;
     }
+    // CraftBukkit end
+
 }
