--- a/net/minecraft/entity/ai/brain/task/FarmerVillagerTask.java
+++ b/net/minecraft/entity/ai/brain/task/FarmerVillagerTask.java
@@ -2,8 +2,10 @@
 
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Lists;
+
 import java.util.List;
 import javax.annotation.Nullable;
+
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
@@ -26,6 +28,7 @@
 import net.minecraft.util.math.Position;
 import net.minecraft.village.VillagerProfession;
 import net.minecraft.world.GameRules;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class FarmerVillagerTask extends Task<VillagerEntity> {
 
@@ -80,8 +83,8 @@
 
     protected void run(ServerWorld serverworld, VillagerEntity villagerentity, long i) {
         if (i > this.nextResponseTime && this.currentTarget != null) {
-            villagerentity.getBrain().remember(MemoryModuleType.LOOK_TARGET, (Object) (new BlockPosLookTarget(this.currentTarget)));
-            villagerentity.getBrain().remember(MemoryModuleType.WALK_TARGET, (Object) (new WalkTarget(new BlockPosLookTarget(this.currentTarget), 0.5F, 1)));
+            villagerentity.getBrain().remember(MemoryModuleType.LOOK_TARGET, (new BlockPosLookTarget(this.currentTarget))); // CraftBukkit - decompile error
+            villagerentity.getBrain().remember(MemoryModuleType.WALK_TARGET, (new WalkTarget(new BlockPosLookTarget(this.currentTarget), 0.5F, 1))); // CraftBukkit - decompile error
         }
 
     }
@@ -101,7 +104,11 @@
                 Block block1 = serverworld.getBlockState(this.currentTarget.down()).getBlock();
 
                 if (block instanceof CropBlock && ((CropBlock) block).isMature(blockstate)) {
-                    serverworld.breakBlock(this.currentTarget, true, villagerentity);
+                    // CraftBukkit start
+                    if (!CraftEventFactory.callEntityChangeBlockEvent(villagerentity, this.currentTarget, Blocks.AIR.getDefaultState()).isCancelled()) {
+                        serverworld.breakBlock(this.currentTarget, true, villagerentity);
+                    }
+                    // CraftBukkit end
                 }
 
                 if (blockstate.isAir() && block1 instanceof FarmlandBlock && villagerentity.hasSeedToPlant()) {
@@ -112,19 +119,29 @@
                         boolean flag = false;
 
                         if (!itemstack.isEmpty()) {
+                            // CraftBukkit start
+                            Block planted = null;
+
                             if (itemstack.getItem() == Items.WHEAT_SEEDS) {
-                                serverworld.setBlockState(this.currentTarget, Blocks.WHEAT.getDefaultState(), 3);
+                                planted = Blocks.WHEAT;
                                 flag = true;
                             } else if (itemstack.getItem() == Items.POTATO) {
-                                serverworld.setBlockState(this.currentTarget, Blocks.POTATOES.getDefaultState(), 3);
+                                planted = Blocks.POTATOES;
                                 flag = true;
                             } else if (itemstack.getItem() == Items.CARROT) {
-                                serverworld.setBlockState(this.currentTarget, Blocks.CARROTS.getDefaultState(), 3);
+                                planted = Blocks.CARROTS;
                                 flag = true;
                             } else if (itemstack.getItem() == Items.BEETROOT_SEEDS) {
-                                serverworld.setBlockState(this.currentTarget, Blocks.BEETROOTS.getDefaultState(), 3);
+                                planted = Blocks.BEETROOTS;
                                 flag = true;
                             }
+
+                            if (planted != null && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(villagerentity, this.currentTarget, planted.getDefaultState()).isCancelled()) {
+                                serverworld.setBlockState(this.currentTarget, planted.getDefaultState(), 3);
+                            } else {
+                                flag = false;
+                            }
+                            // CraftBukkit end
                         }
 
                         if (flag) {
@@ -143,8 +160,8 @@
                     this.currentTarget = this.chooseRandomTarget(serverworld);
                     if (this.currentTarget != null) {
                         this.nextResponseTime = i + 20L;
-                        villagerentity.getBrain().remember(MemoryModuleType.WALK_TARGET, (Object) (new WalkTarget(new BlockPosLookTarget(this.currentTarget), 0.5F, 1)));
-                        villagerentity.getBrain().remember(MemoryModuleType.LOOK_TARGET, (Object) (new BlockPosLookTarget(this.currentTarget)));
+                        villagerentity.getBrain().remember(MemoryModuleType.WALK_TARGET, (new WalkTarget(new BlockPosLookTarget(this.currentTarget), 0.5F, 1))); // CraftBukkit - decompile error
+                        villagerentity.getBrain().remember(MemoryModuleType.LOOK_TARGET, (new BlockPosLookTarget(this.currentTarget))); // CraftBukkit - decompile error
                     }
                 }
             }
@@ -157,6 +174,7 @@
         return this.ticksRan < 200;
     }
 
+    /*
     protected boolean shouldRun(ServerWorld serverworld, LivingEntity livingentity) {
         return this.shouldRun(serverworld, (VillagerEntity) livingentity);
     }
@@ -176,4 +194,5 @@
     protected void run(ServerWorld serverworld, LivingEntity livingentity, long i) {
         this.run(serverworld, (VillagerEntity) livingentity, i);
     }
+    */ // Loom - decompile error
 }
