--- a/net/minecraft/entity/ai/brain/task/LoseJobOnSiteLossTask.java
+++ b/net/minecraft/entity/ai/brain/task/LoseJobOnSiteLossTask.java
@@ -8,6 +8,9 @@
 import net.minecraft.server.world.ServerWorld;
 import net.minecraft.village.VillagerData;
 import net.minecraft.village.VillagerProfession;
+import org.bukkit.craftbukkit.entity.CraftVillager;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.VillagerCareerChangeEvent;
 
 public class LoseJobOnSiteLossTask extends Task<VillagerEntity> {
 
@@ -22,7 +25,14 @@
     }
 
     protected void run(ServerWorld serverworld, VillagerEntity villagerentity, long i) {
-        villagerentity.setVillagerData(villagerentity.getVillagerData().withProfession(VillagerProfession.NONE));
+        // CraftBukkit start
+        VillagerCareerChangeEvent event = CraftEventFactory.callVillagerCareerChangeEvent(villagerentity, CraftVillager.nmsToBukkitProfession(VillagerProfession.NONE), VillagerCareerChangeEvent.ChangeReason.EMPLOYED);
+        if (event.isCancelled()) {
+            return;
+        }
+
+        villagerentity.setVillagerData(villagerentity.getVillagerData().withProfession(CraftVillager.bukkitToNmsProfession(event.getProfession())));
+        // CraftBukkit end
         villagerentity.reinitializeBrain(serverworld);
     }
 
