--- a/net/minecraft/entity/ai/brain/task/VillagerBreedTask.java
+++ b/net/minecraft/entity/ai/brain/task/VillagerBreedTask.java
@@ -1,7 +1,9 @@
 package net.minecraft.entity.ai.brain.task;
 
 import com.google.common.collect.ImmutableMap;
+
 import java.util.Optional;
+
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
 import net.minecraft.entity.LivingEntity;
@@ -16,6 +18,8 @@
 import net.minecraft.util.dynamic.GlobalPos;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.poi.PointOfInterestType;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public class VillagerBreedTask extends Task<VillagerEntity> {
 
@@ -108,6 +112,12 @@
     private Optional<VillagerEntity> createChild(VillagerEntity villagerentity, VillagerEntity villagerentity1) {
         VillagerEntity villagerentity2 = villagerentity.createChild(villagerentity1);
 
+        // CraftBukkit start - call EntityBreedEvent
+        if (CraftEventFactory.callEntityBreedEvent(villagerentity2, villagerentity, villagerentity1, null, null, 0).isCancelled()) {
+            return Optional.empty();
+        }
+        // CraftBukkit end
+
         if (villagerentity2 == null) {
             return Optional.empty();
         } else {
@@ -115,7 +125,7 @@
             villagerentity1.setBreedingAge(6000);
             villagerentity2.setBreedingAge(-24000);
             villagerentity2.refreshPositionAndAngles(villagerentity.getX(), villagerentity.getY(), villagerentity.getZ(), 0.0F, 0.0F);
-            villagerentity.world.spawnEntity(villagerentity2);
+            villagerentity.world.spawnEntity(villagerentity2, CreatureSpawnEvent.SpawnReason.BREEDING); // CraftBukkit - added SpawnReason
             villagerentity.world.sendEntityStatus(villagerentity2, (byte) 12);
             return Optional.of(villagerentity2);
         }
@@ -124,7 +134,7 @@
     private void setChildHome(ServerWorld serverworld, VillagerEntity villagerentity, BlockPos blockpos) {
         GlobalPos globalpos = GlobalPos.create(serverworld.getRegistryKey(), blockpos);
 
-        villagerentity.getBrain().remember(MemoryModuleType.HOME, (Object) globalpos);
+        villagerentity.getBrain().remember(MemoryModuleType.HOME, globalpos); // CraftBukkit - decompile error
     }
 
     protected boolean shouldRun(ServerWorld serverworld, LivingEntity livingentity) {
