--- a/net/minecraft/world/ChunkRegion.java
+++ b/net/minecraft/world/ChunkRegion.java
@@ -247,6 +247,12 @@
     }
 
     public boolean spawnEntity(Entity entity) {
+        // CraftBukkit start
+        return spawnEntity(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    public boolean spawnEntity(Entity entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+        // CraftBukkit end
         int i = MathHelper.floor(entity.getX() / 16.0D);
         int j = MathHelper.floor(entity.getZ() / 16.0D);
 
