--- a/net/minecraft/world/CommandBlockExecutor.java
+++ b/net/minecraft/world/CommandBlockExecutor.java
@@ -17,6 +17,7 @@
 import net.minecraft.util.crash.CrashException;
 import net.minecraft.util.crash.CrashReport;
 import net.minecraft.util.crash.CrashReportSection;
+import org.bukkit.command.CommandSender;
 
 public abstract class CommandBlockExecutor implements CommandOutput {
 
@@ -30,6 +31,10 @@
     private Text lastOutput;
     private String command = "";
     private Text customName;
+    // CraftBukkit start
+    @Override
+    public abstract CommandSender getBukkitSender(ServerCommandSource wrapper);
+    // CraftBukkit end
 
     public CommandBlockExecutor() {
         this.customName = CommandBlockExecutor.DEFAULT_NAME;
@@ -126,7 +131,7 @@
 
                         });
 
-                        minecraftserver.getCommandManager().execute(servercommandsource, this.command);
+                        minecraftserver.getCommandManager().execute(servercommandsource, this.command); // CraftBukkit
                     } catch (Throwable throwable) {
                         CrashReport crashreport = CrashReport.create(throwable, "Executing command block");
                         CrashReportSection crashreportsection = crashreport.addElement("Command to be executed");
