--- a/net/minecraft/network/packet/c2s/play/ClientCommandC2SPacket.java
+++ b/net/minecraft/network/packet/c2s/play/ClientCommandC2SPacket.java
@@ -38,9 +38,7 @@
         return this.mountJumpHeight;
     }
 
-    public void apply(PacketListener packetlistener) {
-        this.apply((ServerPlayPacketListener) packetlistener);
-    }
+    // Loom - decompile errors
 
     public static enum Mode {
 
