--- a/net/minecraft/entity/ai/goal/FollowOwnerGoal.java
+++ b/net/minecraft/entity/ai/goal/FollowOwnerGoal.java
@@ -14,6 +14,13 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.WorldView;
 
+// CraftBukkit start
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.event.entity.EntityTeleportEvent;
+// CraftBukkit end
+
 public class FollowOwnerGoal extends Goal {
 
     private final TameableEntity tameable;
@@ -111,7 +118,19 @@
         } else if (!this.canTeleportTo(new BlockPos(i, j, k))) {
             return false;
         } else {
-            this.tameable.refreshPositionAndAngles((double) i + 0.5D, (double) j, (double) k + 0.5D, this.tameable.yaw, this.tameable.pitch);
+            // CraftBukkit start
+            CraftEntity entity = this.tameable.getBukkitEntity();
+            Location to = new Location(entity.getWorld(), (double) i + 0.5D, (double) j, (double) k + 0.5D, this.tameable.yaw, this.tameable.pitch);
+            EntityTeleportEvent event = new EntityTeleportEvent(entity, entity.getLocation(), to);
+            Bukkit.getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return false;
+            }
+            to = event.getTo();
+
+            this.tameable.refreshPositionAndAngles(to.getX(), to.getY(), to.getZ(), to.getYaw(), to.getPitch());
+            // CraftBukkit end
+
             this.navigation.stop();
             return true;
         }
