--- a/net/minecraft/util/UserCache.java
+++ b/net/minecraft/util/UserCache.java
@@ -46,7 +46,7 @@
     private static boolean useRemote;
     private final Map<String, UserCache.Entry> byName = Maps.newHashMap();
     private final Map<UUID, UserCache.Entry> byUuid = Maps.newHashMap();
-    private final Deque<GameProfile> byAccessTime = Lists.newLinkedList();
+    private final Deque<GameProfile> byAccessTime = new java.util.concurrent.LinkedBlockingDeque<GameProfile>(); // CraftBukkit
     private final GameProfileRepository profileRepository;
     protected final Gson gson;
     private final File cacheFile;
@@ -319,16 +319,6 @@
             }
         }
 
-        public JsonElement serialize(Object object, Type type, JsonSerializationContext jsonserializationcontext) {
-            return this.serialize((UserCache.Entry) object, type, jsonserializationcontext);
-        }
-
-        public Object deserialize(JsonElement jsonelement, Type type, JsonDeserializationContext jsondeserializationcontext) throws JsonParseException {
-            return this.deserialize(jsonelement, type, jsondeserializationcontext);
-        }
-
-        JsonConverter(Object object) {
-            this();
-        }
+        // Loom - decompile error
     }
 }
