--- a/net/minecraft/server/command/EffectCommand.java
+++ b/net/minecraft/server/command/EffectCommand.java
@@ -19,6 +19,7 @@
 import net.minecraft.entity.effect.StatusEffect;
 import net.minecraft.entity.effect.StatusEffectInstance;
 import net.minecraft.text.TranslatableText;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
 
 public class EffectCommand {
 
@@ -70,7 +71,7 @@
             if (entity instanceof LivingEntity) {
                 StatusEffectInstance statuseffectinstance = new StatusEffectInstance(statuseffect, k, i, false, flag);
 
-                if (((LivingEntity) entity).addStatusEffect(statuseffectinstance)) {
+                if (((LivingEntity) entity).addStatusEffect(statuseffectinstance, EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
                     ++j;
                 }
             }
@@ -96,7 +97,7 @@
         while (iterator.hasNext()) {
             Entity entity = (Entity) iterator.next();
 
-            if (entity instanceof LivingEntity && ((LivingEntity) entity).clearStatusEffects()) {
+            if (entity instanceof LivingEntity && ((LivingEntity) entity).clearStatusEffects(EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
                 ++i;
             }
         }
@@ -121,7 +122,7 @@
         while (iterator.hasNext()) {
             Entity entity = (Entity) iterator.next();
 
-            if (entity instanceof LivingEntity && ((LivingEntity) entity).removeStatusEffect(statuseffect)) {
+            if (entity instanceof LivingEntity && ((LivingEntity) entity).removeStatusEffect(statuseffect, EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
                 ++i;
             }
         }
