--- a/net/minecraft/block/ChestBlock.java
+++ b/net/minecraft/block/ChestBlock.java
@@ -272,7 +272,7 @@
     }
 
     public DoubleBlockProperties.PropertySource<? extends ChestBlockEntity> a(BlockState blockstate, World world, BlockPos blockpos, boolean flag) {
-        BiPredicate bipredicate;
+        BiPredicate<WorldAccess, BlockPos> bipredicate; // CraftBukkit - decompile error
 
         if (flag) {
             bipredicate = (worldaccess, blockpos1) -> {
