--- a/net/minecraft/block/ChorusFlowerBlock.java
+++ b/net/minecraft/block/ChorusFlowerBlock.java
@@ -15,6 +15,7 @@
 import net.minecraft.world.World;
 import net.minecraft.world.WorldAccess;
 import net.minecraft.world.WorldView;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class ChorusFlowerBlock extends Block {
 
@@ -77,8 +78,12 @@
                 }
 
                 if (flag && isSurroundedByAir(serverworld, blockpos1, (Direction) null) && serverworld.isAir(blockpos.up(2))) {
-                    serverworld.setBlockState(blockpos, this.plantBlock.withConnectionProperties(serverworld, blockpos), 2);
-                    this.grow(serverworld, blockpos1, i);
+                    // CraftBukkit start - add event
+                    if (CraftEventFactory.handleBlockSpreadEvent(serverworld, blockpos, blockpos1, this.getDefaultState().with(AGE, i), 2)) {
+                        serverworld.setBlockState(blockpos, this.plantBlock.withConnectionProperties(serverworld, blockpos), 2);
+                        this.grow(serverworld, blockpos1, i);
+                    }
+                    // CraftBukkit end
                 } else if (i < 4) {
                     j = random.nextInt(4);
                     if (flag1) {
@@ -92,18 +97,30 @@
                         BlockPos blockpos2 = blockpos.offset(direction);
 
                         if (serverworld.isAir(blockpos2) && serverworld.isAir(blockpos2.down()) && isSurroundedByAir(serverworld, blockpos2, direction.getOpposite())) {
-                            this.grow(serverworld, blockpos2, i + 1);
-                            flag2 = true;
+                            // CraftBukkit start - add event
+                            if (CraftEventFactory.handleBlockSpreadEvent(serverworld, blockpos, blockpos2, this.getDefaultState().with(AGE, i + 1), 2)) {
+                                this.grow(serverworld, blockpos2, i + 1);
+                                flag2 = true;
+                            }
+                            // CraftBukkit end
                         }
                     }
 
                     if (flag2) {
                         serverworld.setBlockState(blockpos, this.plantBlock.withConnectionProperties(serverworld, blockpos), 2);
                     } else {
-                        this.die(serverworld, blockpos);
+                        // CraftBukkit - add event
+                        if (CraftEventFactory.handleBlockGrowEvent(serverworld, blockpos, this.getDefaultState().with(AGE, 5), 2)) {
+                            this.die(serverworld, blockpos);
+                        }
+                        // CraftBukkit end
                     }
                 } else {
-                    this.die(serverworld, blockpos);
+                    // CraftBukkit - add event
+                    if (CraftEventFactory.handleBlockGrowEvent(serverworld, blockpos, this.getDefaultState().with(AGE, 5), 2)) {
+                        this.die(serverworld, blockpos);
+                    }
+                    // CraftBukkit end
                 }
 
             }
