--- a/net/minecraft/block/AbstractPressurePlateBlock.java
+++ b/net/minecraft/block/AbstractPressurePlateBlock.java
@@ -12,6 +12,8 @@
 import net.minecraft.world.World;
 import net.minecraft.world.WorldAccess;
 import net.minecraft.world.WorldView;
+import org.bukkit.Bukkit;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public abstract class AbstractPressurePlateBlock extends Block {
 
@@ -70,6 +72,18 @@
         boolean flag = i > 0;
         boolean flag1 = j > 0;
 
+        // CraftBukkit start - Interact Pressure Plate
+        org.bukkit.World bworld = world.getCraftWorld();
+
+        if (flag != flag1) {
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bworld.getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ()), i, j);
+            Bukkit.getPluginManager().callEvent(eventRedstone);
+
+            flag1 = eventRedstone.getNewCurrent() > 0;
+            j = eventRedstone.getNewCurrent();
+        }
+        // CraftBukkit end
+
         if (i != j) {
             BlockState blockstate1 = this.setRedstoneOutput(blockstate, j);
 
