--- a/net/minecraft/block/CampfireBlock.java
+++ b/net/minecraft/block/CampfireBlock.java
@@ -42,6 +42,7 @@
 import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
 import net.minecraft.world.WorldAccess;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class CampfireBlock extends BlockWithEntity implements Waterloggable {
 
@@ -174,6 +175,11 @@
             if (flag && !(Boolean) blockstate.get(CampfireBlock.LIT) && !(Boolean) blockstate.get(CampfireBlock.WATERLOGGED)) {
                 BlockPos blockpos = blockhitresult.getBlockPos();
 
+                // CraftBukkit start
+                if (CraftEventFactory.callBlockIgniteEvent(world, blockpos, entity).isCancelled()) {
+                    return;
+                }
+                // CraftBukkit end
                 world.setBlockState(blockpos, (BlockState) blockstate.with(Properties.LIT, true), 11);
             }
         }
