--- a/net/minecraft/block/CoralBlock.java
+++ b/net/minecraft/block/CoralBlock.java
@@ -9,6 +9,7 @@
 import net.minecraft.world.BlockView;
 import net.minecraft.world.World;
 import net.minecraft.world.WorldAccess;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class CoralBlock extends CoralParentBlock {
 
@@ -26,7 +27,12 @@
 
     public void scheduledTick(BlockState blockstate, ServerWorld serverworld, BlockPos blockpos, Random random) {
         if (!isInWater(blockstate, serverworld, blockpos)) {
-            serverworld.setBlockState(blockpos, (BlockState) this.deadCoralBlock.getDefaultState().with(CoralBlock.WATERLOGGED, false), 2);
+            // CraftBukkit start
+            if (CraftEventFactory.callBlockFadeEvent(serverworld, blockpos, this.deadCoralBlock.getDefaultState().with(WATERLOGGED, false)).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
+            serverworld.setBlockState(blockpos, this.deadCoralBlock.getDefaultState().with(CoralBlock.WATERLOGGED, false), 2);
         }
 
     }
