--- a/net/minecraft/block/CoralWallFanBlock.java
+++ b/net/minecraft/block/CoralWallFanBlock.java
@@ -7,6 +7,7 @@
 import net.minecraft.util.math.Direction;
 import net.minecraft.world.World;
 import net.minecraft.world.WorldAccess;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class CoralWallFanBlock extends DeadCoralWallFanBlock {
 
@@ -23,7 +24,12 @@
 
     public void scheduledTick(BlockState blockstate, ServerWorld serverworld, BlockPos blockpos, Random random) {
         if (!isInWater(blockstate, serverworld, blockpos)) {
-            serverworld.setBlockState(blockpos, (BlockState) ((BlockState) this.deadCoralBlock.getDefaultState().with(CoralWallFanBlock.WATERLOGGED, false)).with(CoralWallFanBlock.FACING, blockstate.get(CoralWallFanBlock.FACING)), 2);
+            // CraftBukkit start
+            if (CraftEventFactory.callBlockFadeEvent(serverworld, blockpos, this.deadCoralBlock.getDefaultState().with(WATERLOGGED, false).with(FACING, blockstate.get(FACING))).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
+            serverworld.setBlockState(blockpos, this.deadCoralBlock.getDefaultState().with(CoralWallFanBlock.WATERLOGGED, false).with(CoralWallFanBlock.FACING, blockstate.get(CoralWallFanBlock.FACING)), 2);
         }
 
     }
