--- a/net/minecraft/block/CropBlock.java
+++ b/net/minecraft/block/CropBlock.java
@@ -14,6 +14,7 @@
 import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
 import net.minecraft.world.WorldView;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class CropBlock extends PlantBlock implements Fertilizable {
 
@@ -65,7 +66,7 @@
                 float f = getAvailableMoisture(this, serverworld, blockpos);
 
                 if (random.nextInt((int) (25.0F / f) + 1) == 0) {
-                    serverworld.setBlockState(blockpos, this.withAge(i + 1), 2);
+                    CraftEventFactory.handleBlockGrowEvent(serverworld, blockpos, this.withAge(i + 1), 2); // CraftBukkit
                 }
             }
         }
@@ -80,7 +81,7 @@
             i = j;
         }
 
-        world.setBlockState(blockpos, this.withAge(i), 2);
+        CraftEventFactory.handleBlockGrowEvent(world, blockpos, this.withAge(i), 2); // CraftBukkit
     }
 
     protected int getGrowthAmount(World world) {
@@ -136,7 +137,7 @@
     }
 
     public void onEntityCollision(BlockState blockstate, World world, BlockPos blockpos, Entity entity) {
-        if (entity instanceof RavagerEntity && world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING)) {
+        if (entity instanceof RavagerEntity && !CraftEventFactory.callEntityChangeBlockEvent(entity, blockpos, Blocks.AIR.getDefaultState(), !world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING)).isCancelled()) { // CraftBukkit
             world.breakBlock(blockpos, true, entity);
         }
 
