--- a/net/minecraft/block/DoorBlock.java
+++ b/net/minecraft/block/DoorBlock.java
@@ -27,6 +27,8 @@
 import net.minecraft.world.World;
 import net.minecraft.world.WorldAccess;
 import net.minecraft.world.WorldView;
+import org.bukkit.Bukkit;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class DoorBlock extends Block {
 
@@ -173,9 +175,26 @@
     }
 
     public void neighborUpdate(BlockState blockstate, World world, BlockPos blockpos, Block block, BlockPos blockpos1, boolean flag) {
-        boolean flag1 = world.isReceivingRedstonePower(blockpos) || world.isReceivingRedstonePower(blockpos.offset(blockstate.get(DoorBlock.HALF) == DoubleBlockHalf.LOWER ? Direction.UP : Direction.DOWN));
+        //boolean flag1 = world.isReceivingRedstonePower(blockpos) || world.isReceivingRedstonePower(blockpos.offset(blockstate.get(DoorBlock.HALF) == DoubleBlockHalf.LOWER ? Direction.UP : Direction.DOWN));
 
-        if (block != this && flag1 != (Boolean) blockstate.get(DoorBlock.POWERED)) {
+        // CraftBukkit start
+        BlockPos otherHalf = blockpos.offset(blockstate.get(HALF) == DoubleBlockHalf.LOWER ? Direction.UP : Direction.DOWN);
+
+        org.bukkit.World bworld = world.getCraftWorld();
+        org.bukkit.block.Block bukkitBlock = bworld.getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ());
+        org.bukkit.block.Block blockTop = bworld.getBlockAt(otherHalf.getX(), otherHalf.getY(), otherHalf.getZ());
+
+        int power = bukkitBlock.getBlockPower();
+        int powerTop = blockTop.getBlockPower();
+        if (powerTop > power) power = powerTop;
+        int oldPower = blockstate.get(POWERED) ? 15 : 0;
+
+        if (oldPower == 0 ^ power == 0) {
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bukkitBlock, oldPower, power);
+            Bukkit.getPluginManager().callEvent(eventRedstone);
+
+            boolean flag1 = eventRedstone.getNewCurrent() > 0;
+            // CraftBukkit end
             if (flag1 != (Boolean) blockstate.get(DoorBlock.OPEN)) {
                 this.playOpenCloseSound(world, blockpos, flag1);
             }
