--- a/net/minecraft/block/DragonEggBlock.java
+++ b/net/minecraft/block/DragonEggBlock.java
@@ -11,6 +11,8 @@
 import net.minecraft.util.shape.VoxelShape;
 import net.minecraft.world.BlockView;
 import net.minecraft.world.World;
+import org.bukkit.Bukkit;
+import org.bukkit.event.block.BlockFromToEvent;
 
 public class DragonEggBlock extends FallingBlock {
 
@@ -38,6 +40,18 @@
             BlockPos blockpos1 = blockpos.add(world.random.nextInt(16) - world.random.nextInt(16), world.random.nextInt(8) - world.random.nextInt(8), world.random.nextInt(16) - world.random.nextInt(16));
 
             if (world.getBlockState(blockpos1).isAir()) {
+                // CraftBukkit start
+                org.bukkit.block.Block from = world.getCraftWorld().getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ());
+                org.bukkit.block.Block to = world.getCraftWorld().getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ());
+                BlockFromToEvent event = new BlockFromToEvent(from, to);
+                Bukkit.getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return;
+                }
+
+                blockpos1 = new BlockPos(event.getToBlock().getX(), event.getToBlock().getY(), event.getToBlock().getZ());
+                // CraftBukkit end
                 if (world.isClient) {
                     for (int j = 0; j < 128; ++j) {
                         double d0 = world.random.nextDouble();
