--- a/net/minecraft/block/SpreadableBlock.java
+++ b/net/minecraft/block/SpreadableBlock.java
@@ -7,6 +7,7 @@
 import net.minecraft.util.math.Direction;
 import net.minecraft.world.WorldView;
 import net.minecraft.world.chunk.light.ChunkLightProvider;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public abstract class SpreadableBlock extends SnowyBlock {
 
@@ -37,6 +38,11 @@
 
     public void randomTick(BlockState blockstate, ServerWorld serverworld, BlockPos blockpos, Random random) {
         if (!canSurvive(blockstate, serverworld, blockpos)) {
+            // CraftBukkit start
+            if (CraftEventFactory.callBlockFadeEvent(serverworld, blockpos, Blocks.DIRT.getDefaultState()).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
             serverworld.setBlockState(blockpos, Blocks.DIRT.getDefaultState());
         } else {
             if (serverworld.getLightLevel(blockpos.up()) >= 9) {
@@ -46,7 +52,7 @@
                     BlockPos blockpos1 = blockpos.add(random.nextInt(3) - 1, random.nextInt(5) - 3, random.nextInt(3) - 1);
 
                     if (serverworld.getBlockState(blockpos1).isOf(Blocks.DIRT) && canSpread(blockstate1, serverworld, blockpos1)) {
-                        serverworld.setBlockState(blockpos1, (BlockState) blockstate1.with(SpreadableBlock.SNOWY, serverworld.getBlockState(blockpos1.up()).isOf(Blocks.SNOW)));
+                        CraftEventFactory.handleBlockSpreadEvent(serverworld, blockpos, blockpos1, blockstate1.with(SNOWY, serverworld.getBlockState(blockpos1.up()).isOf(Blocks.SNOW))); // CraftBukkit
                     }
                 }
             }
