--- a/net/minecraft/nbt/CompoundTag.java
+++ b/net/minecraft/nbt/CompoundTag.java
@@ -74,13 +74,11 @@
             return "TAG_Compound";
         }
 
-        public Tag read(DataInput datainput, int i, PositionTracker positiontracker) throws IOException {
-            return this.read(datainput, i, positiontracker);
-        }
+        // Loom - decompile error
     };
     private final Map<String, Tag> tags;
 
-    protected CompoundTag(Map<String, Tag> map) {
+    public CompoundTag(Map<String, Tag> map) { // Loom - public access
         this.tags = map;
     }
 
@@ -526,11 +524,9 @@
         }
     }
 
-    protected Map<String, Tag> method_29143() {
+    public Map<String, Tag> method_29143() { // Loom - public access
         return Collections.unmodifiableMap(this.tags);
     }
 
-    public Tag copy() {
-        return this.copy();
-    }
+    // Loom - decompile error
 }
