--- a/net/minecraft/world/storage/RegionBasedStorage.java
+++ b/net/minecraft/world/storage/RegionBasedStorage.java
@@ -23,7 +23,7 @@
         this.desync = flag;
     }
 
-    private RegionFile getRegionFile(ChunkPos chunkpos) throws IOException {
+    private RegionFile getRegionFile(ChunkPos chunkpos, boolean existingOnly) throws IOException { // CraftBukkit
         long i = ChunkPos.toLong(chunkpos.getRegionX(), chunkpos.getRegionZ());
         RegionFile regionfile = (RegionFile) this.cachedRegionFiles.getAndMoveToFirst(i);
 
@@ -39,6 +39,7 @@
             }
 
             File file = new File(this.directory, "r." + chunkpos.getRegionX() + "." + chunkpos.getRegionZ() + ".mca");
+            if (existingOnly && !file.exists()) return null; // CraftBukkit
             RegionFile regionfile1 = new RegionFile(file, this.directory, this.desync);
 
             this.cachedRegionFiles.putAndMoveToFirst(i, regionfile1);
@@ -48,7 +49,12 @@
 
     @Nullable
     public CompoundTag getTagAt(ChunkPos chunkpos) throws IOException {
-        RegionFile regionfile = this.getRegionFile(chunkpos);
+        // CraftBukkit start - SPIGOT-5680: There's no good reason to preemptively create files on read, save that for writing
+        RegionFile regionfile = this.getRegionFile(chunkpos, true);
+        if (regionfile == null) {
+            return null;
+        }
+        // CraftBukkit end
         DataInputStream datainputstream = regionfile.getChunkInputStream(chunkpos);
         Throwable throwable = null;
 
@@ -83,7 +89,7 @@
     }
 
     protected void write(ChunkPos chunkpos, CompoundTag compoundtag) throws IOException {
-        RegionFile regionfile = this.getRegionFile(chunkpos);
+        RegionFile regionfile = this.getRegionFile(chunkpos, false); // CraftBukkit
         DataOutputStream dataoutputstream = regionfile.getChunkOutputStream(chunkpos);
         Throwable throwable = null;
 
